packages/network/src/workers/syncUtils.ts

```diff
@@ -485,21 +491,58 @@ export function createTransformWorldEventsFromStream(decode: ReturnType<typeof c
   };
 }
 
+function mapTransactionHashesToNetworkComponentUpdates(events: NetworkComponentUpdate[]) {
+  return events.reduce((acc, event) => {
+    if (["worker", "cache"].includes(event.txHash)) return acc;
+
+    if (!acc[event.txHash]) acc[event.txHash] = [];
+
+    acc[event.txHash].push(event);
+
+    return acc;
+  }, {} as { [key: string]: NetworkComponentUpdate[] });
+}
+
+function parseSystemCallTransactionFromStreamNetworkComponentUpdate(event: NetworkComponentUpdate) {
+  if (!event.txMetadata) return null;
+
+  const { to, data, value } = event.txMetadata;
+
+  return {
+    to,
+    data: BigNumber.from(data).toHexString(),
+    value: BigNumber.from(value),
+    hash: event.txHash,
+  } as SystemCallTransaction;
+}
+
+export function parseSystemCallsFromStreamEvents(events: NetworkComponentUpdate[]) {
+  const systemCalls: SystemCall[] = [];
+  const transactionHashToEvents = mapTransactionHashesToNetworkComponentUpdates(events);
+
+  for (const txHash of Object.keys(transactionHashToEvents)) {
+    // All ECS events include the information needed to parse the SysytemCallTransasction out, so it doesn't
+    // matter which one we take here.
+    const tx = parseSystemCallTransactionFromStreamNetworkComponentUpdate(transactionHashToEvents[txHash][0]);
+    if (!tx) continue;
+
+    systemCalls.push({
+      type: NetworkEvents.SystemCall,
+      tx,
+      updates: transactionHashToEvents[tx.hash],
+    });
+  }
+
+  return systemCalls;
+}
+
 export function createFetchSystemCallsFromEvents(provider: JsonRpcProvider) {
   const { fetchBlock, clearBlock } = createBlockCache(provider);
   const fetchSystemCallData = createFetchSystemCallData(fetchBlock);
 
   return async (events: NetworkComponentUpdate[], blockNumber: number) => {
     const systemCalls: SystemCall[] = [];
-    const transactionHashToEvents = events.reduce((acc, event) => {
-      if (["worker", "cache"].includes(event.txHash)) return acc;
-
-      if (!acc[event.txHash]) acc[event.txHash] = [];
-
-      acc[event.txHash].push(event);
-
-      return acc;
-    }, {} as { [key: string]: NetworkComponentUpdate[] });
+    const transactionHashToEvents = mapTransactionHashesToNetworkComponentUpdates(events);
```

Comment:
```suggestion
    const transactionHashToEvents = groupByTxHash(events);
```
(if above suggestion is accepted)